src/Microsoft.OData.Core/UriParser/Parsers/FunctionOverloadResolver.cs

```diff
@@ -164,45 +164,43 @@ internal static bool ResolveOperationFromList(string identifier, IEnumerable<str
-            {
-                candidateMatchingOperations.EnsureOperationsBoundWithBindingParameter();
-            }
+            IList<IEdmOperation> candidatesMatchingOperations = operationsFromModel.FilterOperationCandidatesBasedOnParameterList(bindingType, parameterNames, resolver.EnableCaseInsensitive, out foundActionsWhenLookingForFunctions);
 
-            // If the number of parameters > 0 then this has to be a function as actions can't have parameters on the uri only in the payload. Filter further by parameters in this case, otherwise don't.
-            if (hasParameters)
-            {
-                // can only be a function as only functions have parameters on the uri.
-                candidateMatchingOperations = candidateMatchingOperations.RemoveActions(out foundActionsWhenLookingForFunctions)
-                    .FilterOperationsByParameterNames(parameterNames, resolver.EnableCaseInsensitive);
-            }
-            else if (bindingType != null)
-            {
-                // Filter out functions with more than one parameter. Actions should not be filtered as the parameters are in the payload not the uri
-                candidateMatchingOperations = candidateMatchingOperations.Where(o =>
-                (o.IsFunction() && (o.Parameters.Count() == 1 || o.Parameters.Skip(1).All(p => p is IEdmOptionalParameter))) || o.IsAction());
-            }
-            else
+            // Only filter if there is more than one and its needed.
+            if (candidatesMatchingOperations.Count > 1)
             {
-                // Filter out functions with any parameters
-                candidateMatchingOperations = candidateMatchingOperations.Where(o => (o.IsFunction() && !o.Parameters.Any()) || o.IsAction());
+                candidatesMatchingOperations = candidatesMatchingOperations.FilterBoundOperationsWithSameTypeHierarchyToTypeClosestToBindingType(bindingType) as IList<IEdmOperation>;
             }
 
-            // Only filter if there is more than one and its needed.
-            if (candidateMatchingOperations.Count() > 1)
+            // If any of the things returned are an action, it better be the only thing returned, and there can't be parameters in the URL
+            bool actionExists = false;
+            bool functionExists = false;
+            for (int i = 0; i < candidatesMatchingOperations.Count; i++)
             {
-                candidateMatchingOperations = candidateMatchingOperations.FilterBoundOperationsWithSameTypeHierarchyToTypeClosestToBindingType(bindingType);
+                if (candidatesMatchingOperations[i].IsFunction())
+                {
+                    functionExists = true;
+                }
+
+                if (candidatesMatchingOperations[i].IsAction())
+                {
+                    actionExists = true;
+                }
+
+                if (functionExists && actionExists)
+                {
+                    break;
+                }

```

Comment:
split and move to 185; and 190?

Reply:
Can't move both the exceptions to above as we need to traverse the entire list before deciding if both action and function exist vs only actions exists. 

---
In reply to: [370888279](https://github.com/OData/odata.net/pull/1602#discussion_r370888279) [](ancestors = 370888279)